new_platforms/src/Trusted/optee/cyres_optee.c

```diff
@@ -1,223 +1,224 @@
+    if (oe_result != OE_OK)
+    {
+        if (oe_result == OE_BUFFER_TOO_SMALL)
+            *buf_size = params[0].memref.size;
+        return oe_result;
+    }
 
-Tcps_ReturnStatusCode;
-Tcps_BeginErrorHandling;
-Tcps_FinishErrorHandling;
+    *buf_size = params[0].memref.size;
+    return OE_OK;
 }
 
-oe_result_t 
-ExportCyrepCertChain(
-    Tcps_ConstStringA exportFilePath)
+/* Common PTA routine for buffer retrival commands */
+static TEE_Result get_pta_allocated_buf(
+    uint32_t cmd_id,
+    uint8_t** buf,
+    uint32_t* buf_size)
 {
-    uint32_t pt;
-    TEE_Result teeResult;
-    Tcps_UInt32 certChainBufferSize;
-    char *certChainBufferPEM = NULL;
-    TEE_Param params[TEE_NUM_PARAMS];
-
-Tcps_InitializeStatus(Tcps_Module_Helper_t, "ExportCyrepCertChain");
+    oe_result_t oe_result;
+    uint32_t local_buf_size;
+    uint8_t* local_buf = NULL;
 
-    /* Get the chain size */
-    uStatus = GetCyrepCertChainSize(&certChainBufferSize);
-    Tcps_GotoErrorIfBad(uStatus);
-    oe_assert(certChainBufferSize != 0);
-
-    /* Allocate memory and request the cert chain */
-    certChainBufferPEM = oe_malloc(certChainBufferSize);
-    Tcps_ReturnErrorIfAllocFailed(certChainBufferPEM);
-    memset(certChainBufferPEM, 0, certChainBufferSize);
-
-    pt = TEE_PARAM_TYPES(
-		TEE_PARAM_TYPE_MEMREF_INOUT,
-		TEE_PARAM_TYPE_NONE,
-		TEE_PARAM_TYPE_NONE,
-		TEE_PARAM_TYPE_NONE);
-    memset(params, 0, sizeof(params));
-
-    params[0].memref.buffer = certChainBufferPEM;
-    params[0].memref.size = certChainBufferSize;
-
-    teeResult = CallCyrepPTA(PTA_CYREP_GET_CERT_CHAIN, pt, params);
-    Tcps_GotoErrorIfTrue(teeResult != TEE_SUCCESS, OE_FAILURE);
-
-    /* PEM format, should have a zero terminator */
-    oe_assert(certChainBufferPEM[certChainBufferSize - 1] == 0);
+    /* Get the data size */
+    oe_result = get_pta_buf(cmd_id, NULL, &local_buf_size);
+    if (oe_result != OE_BUFFER_TOO_SMALL)
+        goto done;
 
-    /* Export the cert chain into an untrusted-world file */
-    uStatus = TEE_P_ExportFile(exportFilePath, certChainBufferPEM, certChainBufferSize);
-    Tcps_GotoErrorIfBad(uStatus);
+    oe_assert(local_buf_size != 0);
 
-    if (certChainBufferPEM != NULL)
+    /* Allocate memory locally and request the data */
+    local_buf = oe_malloc(local_buf_size);
+    if (local_buf == NULL)
     {
-        oe_free(certChainBufferPEM);
+        oe_result = OE_OUT_OF_MEMORY;
+        goto done;
     }
+    memset(local_buf, 0, local_buf_size);
 
-Tcps_ReturnStatusCode;
-Tcps_BeginErrorHandling;
-    if (certChainBufferPEM != NULL)
-    {
-        oe_free(certChainBufferPEM);
-    }
-Tcps_FinishErrorHandling;
+    oe_result = get_pta_buf(cmd_id, local_buf, &local_buf_size);
+    if (oe_result != OE_OK)
+        goto done;
+
+    /* Remember to clear the local reference on success */
+    oe_result = OE_OK;
+    *buf = local_buf;
+    local_buf = NULL;
+    *buf_size = local_buf_size;
+
+done:
+    oe_free(local_buf);
```

Comment:
Probably don't want to free local_buf in success case since it gets returned via out parameter.

Reply:
I see you're assigning it to NULL above. A little confusing but OK.

---
In reply to: [235234141](https://github.com/Microsoft/openenclave/pull/1095#discussion_r235234141) [](ancestors = 235234141)

Reply:
Yeah. I can place a comment. The idea with this pattern is to always free the local pointer. If the buffer needs to be returned, the reference should be transferred before the cleanup code.